A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,296 @@
+warning and pass the RPC through. For a valid `override-host` value set it
+in the RPC context as the value of `upstreamAddress` state variable which
+will be used in the response processing of that RPC.
+
+* In case of a valid `override-host` value, pass this value as a
+"pick-argument" to the Load Balancer's pick method as described in [A31][].
+For example,
+
+    * in Java use `CallOptions` to pass this value.
+
+    * in Go pass it via `Context`.
+
+    * In C-core, this will be via the `GRPC_CONTEXT_SERVICE_CONFIG_CALL_DATA`
+      call context element whose API will be extended to allow the
+      filter to add a new call element.
+
+* In the response path of the RPC, if the RPC path does not match the
+  `path` config value of the filter then skip further processing and
+  pass the response through. Otherwise check if `upstreamAddress` is
+  set (in the RPC context of this filter) and get the value. Also get
+  the peer address for the RPC; let’s call it `hostAddress` here. For
+  example, in Java you get the peer address through the
+  `Grpc.TRANSPORT_ATTR_REMOTE_ADDR` attribute of the current
+  ClientCall. If `upstreamAddress` is not set or `upstreamAddress` and
+  `hostAddress` are different, then create a cookie using our Cookie config
+  which has the `name`, `ttl` and `path` values. Set the value of the cookie
+  to be the base64-encoded string value of `hostAddress`. Add this Cookie to
+  the response headers. The logic in pseudo-code is:
+
+```C++
+  if (!upstreamAddress.has_value() || hostAddress != upstreamAddress) {
+    String encoded_address = base64_encode(hostAddress);
+    Create a Cookie from our cookie configuration;
+    Set cookie's value as encoded_address;
+    Add the cookie header to the response;
+  }
+```
+
+### LB Policy for Stateful Session Affinity
+
+After the `CookieBasedStatefulSessionFilter` has passed the `override-host`
+value to the Load Balancer as a "pick-argument", the Load Balancer uses
+this value to route the RPC appropriately. The required logic needs to be
+implemented in a new and separate LB policy at an appropriate place in the
+hierarchy.
+
+RPC load reporting happens in the `“xds_cluster_impl”` policy and we do want
+all RPCs to be included in the load reports. And this policy conditionally
+routes an RPC based on the `override-host` value if present, or else delegates
+to the currently configured LB policy. Hence the new policy needs to be just
+below the `“xds_cluster_impl”` as its child policy. Let's call this new policy
+`“override-host-experimental”`. This policy has the required subchannel and
+RPC routing logic as follows:
+
+* maintain a map of addresses (IP:port) to subchannels - let’s call this map
+  `overrideHostMap`.
+
+* whenever a new subchannel is created (as a result of an endpoint update from
+  the control plane), add a new entry to `overrideHostMap` with the subchannel
+  address(es) as the key and the subchannel as the value.
```

Comment:
by `address` here you mean peer address

Reply:
Yes, the subchannel address (as a property of a subchannel) is the peer address - I could add clarification here.

Reply:
clarification added